src/python/pants/backend/adhoc/run_system_binary_integration_test.py

```diff
@@ -91,3 +91,35 @@ def test_fingerprint(fingerprint: str, passes: bool) -> None:
         else:
             assert result.exit_code != 0
             assert "Could not find a binary with name `bash`" in result.stderr.strip()
+
+
+def test_runnable_dependencies() -> None:
+    sources = {
+        "src/BUILD": dedent(
+            """\
+            system_binary(
+                name="bash",
+                binary_name="bash",
+            )
+
+            adhoc_tool(
+                name="adhoc",
+                runnable=":bash",
+                execution_dependencies=[_runnable(name="bash", address=":bash"),],
```

Comment:
I think I agree with @stuhood that it would make things easier and clearer to have a dedicated field for the runnable dependencies rather than an object wrapper.

Although another option could be to simply detect what kind of target is listed as execution dependency. If it is runnable provide it on the path (likely using its target name), if it is a resource just dump the files in the sandbox if it is packagable build it and dump the artifact etc. We would loose the ability to override name for the executables though.. but I don't think that would be too bad as a start any way. 

However, to me this feels like yet another example where we would like to be able to annotate dependency edges. Would prefer to get that moving before we have a dozen different use-cases all with slightly different implementations for solving this in isolation.

Reply:
> Although another option could be to simply detect what kind of target is listed as execution dependency. If it is runnable provide it on the path (likely using its target name), if it is a resource just dump the files in the sandbox if it is packagable build it and dump the artifact etc. We would loose the ability to override name for the executables though.. but I don't think that would be too bad as a start any way.

Annoyingly, specifying the name is absolutely necessary -- `yarn`'s launcher binary is a javascript source that starts with `#!/usr/bin/env node`.

Similarly, knowing how a dependency is to be used is also important -- the implementation I've offered allows e.g. `pex_binary` targets to be provided as runnable (if you want to run the pex), or as data (if you want the runnable specified by `adhoc_tool` to consume the pex, maybe as an upload process or something). We need some sort of way of disambiguating.

Reply:
> > Although another option could be to simply detect what kind of target is listed as execution dependency. If it is runnable provide it on the path (likely using its target name), if it is a resource just dump the files in the sandbox if it is packagable build it and dump the artifact etc. We would loose the ability to override name for the executables though.. but I don't think that would be too bad as a start any way.
> 
> Annoyingly, specifying the name is absolutely necessary -- `yarn`'s launcher binary is a javascript source that starts with `#!/usr/bin/env node`.
> 

Not saying this was my favourite route; but it would still work, the restriction being that the node runnable be a `system_binary(name="node")`.

> Similarly, knowing how a dependency is to be used is also important -- the implementation I've offered allows e.g. `pex_binary` targets to be provided as runnable (if you want to run the pex), or as data (if you want the runnable specified by `adhoc_tool` to consume the pex, maybe as an upload process or something). We need some sort of way of disambiguating.

Ah, so this highlights my final section that we'd like to annotate the connection somehow to indicate how it should be used. Lacking that we're turning to dedicated fields for specific use cases that becomes very target specific where I would much prefer a more general solution to lean on.

Reply:
> Not saying this was my favourite route; but it would still work, the restriction being that the node runnable be a `system_binary(name="node")`.

This would work, but it could become cumbersome (it would need more docs

> turning to dedicated fields for specific use cases that becomes very target specific where I would much prefer a more general solution to lean on 

Agreed that we'd prefer a more general solution. #18320 suggests that the best way to tackle this would be through scoped dependencies (what you call "annotating the connection", which is effectively what's going on with `_runnable`).

`adhoc_tool` highlights that there's two different axes for dependency scoping: _where_ the dependency is used (e.g. a dependency of the output of the tool vs. a dependency needed to produce the output) and _how_ the dependency is used (e.g. as data for the tool at runtime vs. as an executable the tool can execute). 

We've (temporarily) solved the "where" question by having two separate fields for dependencies, but the downside to this is that we can't make runnable dependencies transitive, because we can only declare runnable dependencies where this field is defined. The result is that build files end up being repetitive (see [the example I've been working through](https://gist.github.com/chrisjrn/42677a04f6631779d2e0002d0acca624)): you need to define and redefine runnable dependencies. Better would be able to describe -- precisely once -- that `yarn` needs a command named `node` on its `PATH` when run, and then anything that needs to depend on running `yarn` will have that underlying `node` binary available. That's not possible without the cooperation of core/plugin authors if we require new fields to describe each context for dependencies.

I'd like to see this achieved with the standard `dependencies` field, so that things like `execution_dependencies` can disappear entirely. In the meantime, experimenting within the `execution_dependencies` field feels like a good way to see if the idea works? 

---

Re. using a custom object vs dicts or strings, I leaned towards custom objects because it allows for automatically validating the metadata for the dependency (in this case, that a `runnable` dependency needs a name for the binary). I am open to other ways of doing this.